### PR TITLE
Add incremental_search config option

### LINKED ISSUE #584: Incremental search
While searching with `/`, the results only appear after hitting enter. Better to highlight as you type, as in luakit.

### PR DESCRIPTION
**Pull request by jupart**
The first commit adds incremental search. The second commit is a kind of quality of life improvement I felt was needed as you can spam 'Text {} not found on page!' warnings a lot with this new feature.

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3201)

edit by @The-Compiler: Closes #584

### CODE REVIEW

File: qutebrowser/mainwindow/statusbar/command.py [hunk 1]
```diff
@@ -53,6 +55,7 @@ class Command(misc.MinimalLineEditMixin, misc.CommandLineEdit):
     update_completion = pyqtSignal()
     show_cmd = pyqtSignal()
     hide_cmd = pyqtSignal()
+    update_search = pyqtSignal()
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/command.py]: No need to add a new signal here - just connect `textChanged` to the function directly.

File: qutebrowser/mainwindow/statusbar/command.py [hunk 2]
```diff
@@ -216,3 +221,13 @@ def sizeHint(self):
             text = 'x'
         width = self.fontMetrics().width(text)
         return QSize(width, height)
+
+    def incrementSearch(self):
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/command.py]: Various things to note here:

- Please add a `@pyqtSlot()` decorator to mark this as a slot (a method connected to the signal)
- The `textChanged` signal also gives you the new text, so you can use `@pyqtSlot(str)` and get a `text` argument, which makes this a bit easier.
- This should be called `increment_search` to follow [pep8](https://www.python.org/dev/peps/pep-0008/)
- However, I'd prever `incremental_search`
- However, this is "private", so it should be `_incremental_search` :wink:

File: qutebrowser/mainwindow/statusbar/command.py [hunk 3]
```diff
@@ -216,3 +221,13 @@ def sizeHint(self):
...
+        if config.val.incremental_search == True:
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/command.py]: No need for the `== True`

File: qutebrowser/mainwindow/statusbar/command.py [hunk 4]
```diff
@@ -216,3 +221,13 @@ def sizeHint(self):
...
+            search_prefixes = {
+                '/': 'search -- ',
+                '?': 'search -r -- ',
+            }
+            if self.prefix() == '/' or self.prefix() == '?':
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/command.py]: `if self.prefix() in '/?':`

File: qutebrowser/config/configdata.yml [hunk 5]
```diff
@@ -43,6 +43,11 @@ ignore_case:
   default: smart
   desc: Find text on a page case-insensitively.
 
+incremental_search:
+  type: Bool
+  default: False
+  desc: Find text on a page incrementally, renewing the search for each typing character.
+
```
Comment (The-Compiler) [inline qutebrowser/config/configdata.yml]: Hmm, I think at this point we should rename `ignore_case` to `search.ignore_case`, and call this `search.incremental`. Please do so in a separate commit though - you'll need to keep `ignore_case` with `renamed: search.ignore_case` as value in here, and also adjust any code using `ignore_case`.

File: qutebrowser/config/configdata.yml [hunk 6]
```diff
@@ -43,6 +43,11 @@ ignore_case:
   default: smart
   desc: Find text on a page case-insensitively.
 
+incremental_search:
+  type: Bool
+  default: False
```
Comment (The-Compiler) [inline qutebrowser/config/configdata.yml]: I think it's fine to default this to True, so people actually discover that it exists :wink:

File: qutebrowser/browser/commands.py [hunk 7]
```diff
@@ -1742,7 +1742,8 @@ def _search_cb(self, found, *, tab, old_scroll_pos, options, text, prev):
             elif going_up and tab.scroller.pos_px().y() > old_scroll_pos.y():
                 message.info("Search hit TOP, continuing at BOTTOM")
         else:
-            message.warning("Text '{}' not found on page!".format(text))
+            message.warning("Text '{}' not found on page!".format(text),
+                    replace=True)
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: Please only indent this by four spaces.

File: qutebrowser/mainwindow/statusbar/command.py [hunk 8]
```diff
@@ -45,6 +46,7 @@ class Command(misc.MinimalLineEditMixin, misc.CommandLineEdit):
         update_completion: Emitted when the completion should be shown/updated.
         show_cmd: Emitted when command input should be shown.
         hide_cmd: Emitted when command input can be hidden.
+        update_search: Emitted when search term could be updated.
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/command.py]: You'll need to remove this again (see below) :wink:

File: qutebrowser/config/configdata.yml [hunk 9]
```diff
@@ -33,7 +33,7 @@ history_gap_interval:
     Items with less time between them are grouped when being displayed in `:history`.
     Use -1 to disable separation.
 
-ignore_case:
+search.ignore_case:
```
Comment (jgkamat) [inline qutebrowser/config/configdata.yml]: I think you want to add a `renamed:` argument so the autoconfig can be migrated properly for this key

**Comment by jgkamat** [comment]
Re: typing fast, submitting a second request to search while one is still running causes that error to be displayed.

I dont know how difficult this is, but maybe we could hold off on submitting the request while the user is typing or throttle the requests to search.

File: qutebrowser/config/configdata.yml [hunk 10]
```diff
@@ -34,6 +34,9 @@ history_gap_interval:
     Use -1 to disable separation.
 
 ignore_case:
+  renamed: search.ignore_case
```
Comment (The-Compiler) [inline qutebrowser/config/configdata.yml]: Various tests also use `ignore_case` to test something and now [fail](https://ci.appveyor.com/project/qutebrowser/qutebrowser/build/master-2361/job/sblwxgtsnmu64o1q), so you'll need to adjust those as well.

File: qutebrowser/mainwindow/statusbar/command.py [hunk 11]
```diff
@@ -216,3 +218,15 @@ def sizeHint(self):
             text = 'x'
         width = self.fontMetrics().width(text)
         return QSize(width, height)
+
+    @pyqtSlot(str)
+    def _incremental_search(self, text):
+        if config.val.search.incremental:
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/command.py]: just do this instead:

```python
if config.val.search.incremental:
    return
```

and dedent the rest of the function by one.

File: qutebrowser/mainwindow/statusbar/command.py [hunk 12]
```diff
@@ -216,3 +218,15 @@ def sizeHint(self):
...
+            search_prefixes = {
+                '/': 'search -- ',
+                '?': 'search -r -- ',
+            }
+
+            # `if text` check is for when user presses <Esc> and `len(text)` is 0
+            if self.prefix() in '/?' and text:
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/command.py]: Hmm, `'' in '/?'` being `True` is a quite subtle difference between that and `'' in ['/', '?']`, I never realized that!

I think it'd be nicer to rewrite this to `if self.prefix() in ['/', '?']:` instead, then with an empty `text`, `self.prefix()` will return `''` and that will be `False`.

**Comment by jupart** [comment]
I have a 'solution' for the timeout that uses QTimer.setSingleShot timers to wait some settable timeout before searching after text was input. But I think the issue lies with Qt search in the tab (as far as I can trace it back). I think the Qt search returns `found = False` when spammed with search requests that are too fast for it to handle. You can test this on `master` by `bind <Ctrl-s> search some_term_on_your_page`.

@The-Compiler, if you'd like I can push the current solution I have, but I don't think it's a good fix for this issue.

**Comment by jupart** [comment]
Maybe a better solution is to simply clear the message after every keypress.

**Comment by The-Compiler** [comment]
Sorry for the delay - some comments about this:

- As for the flickering, this is because qutebrowser clears the search every time before a new search happens:

https://github.com/qutebrowser/qutebrowser/blob/43aa7423abb931ddcbccc51411f6a8a16bf056b5/qutebrowser/browser/commands.py#L1758-L1759

Commenting this out however means that e.g. we start at the previous position in the page when doing `xy<Backspace>` or so. I don't think we can do much about it.

- The "not found" issue should probably be fixed properly, see #2442. Is this something you'd like to take a look at (probably as a separate PR), @jupart?

- Right now, doing `/foo` and pressing enter jumps to the second match. I guess we should just ignore a `/foo` command if incremental search is done?

**Comment by The-Compiler** [comment]
Any update on this, @jupart? Unfortunately I don't think this is usable before fixing #2442, and I probably won't have the time to look at that in the next few months...

**Comment by The-Compiler** [comment]
Now that #2442 is fixed, can you please merge the latest master? Also, those points are still open IIRC:

- Right now, doing `/foo` and pressing enter jumps to the second match. I guess we should just ignore a `/foo` command if incremental search is done?

- Some tests related to [settings](https://travis-ci.org/qutebrowser/qutebrowser/jobs/293660510) and [searching](https://travis-ci.org/qutebrowser/qutebrowser/jobs/293660511) now fail. I haven't really looked into them yet, but it seems kind of weird that they fail.

The latter could probably be solved by just turning off incremental search for the tests.

**Comment by The-Compiler** [comment]
This happens because there's now a dot in the ID which is selected via `:click-element`. Not your fault, I'll try to fix it in `master`.

**Comment by The-Compiler** [comment]
This should be fixed with 5fe91c30cc0eb3b960ef1a256a302cf89996a273 now - sorry for that! Can you please merge the latest master into yours? Let me know if you need help with that!

**Comment by jupart** [comment]
Another unintentional PR close when force-pushing to squash all my "Merge into master" commits. jupart/master should be up to date, but I'm having a hard time reopening this one this time. I need to work on my Github-foo.

File: qutebrowser/mainwindow/statusbar/command.py [hunk 13]
```diff
@@ -26,7 +26,8 @@
 from qutebrowser.commands import cmdexc, cmdutils
 from qutebrowser.misc import cmdhistory, editor
 from qutebrowser.misc import miscwidgets as misc
-from qutebrowser.utils import usertypes, log, objreg, message
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/command.py]: You removed the `message` import (bad merge?), causing stuff to break :wink:

**Comment by The-Compiler** [comment]
Looks good to me now! I still want to play around with it a bit before merging it, and I'll probably do that tomorrow. 👍 

As for the GitHub issues: I'm not sure how this happened, but don't worry too much about the noise, there's a lot going on with this repo anyways! :wink: In the future, I suggest you use a dedicated git branch for your changes though, as this definitely makes it easier to manage stuff. See e.g. [this guide](http://api.coala.io/en/latest/Developers/Newcomers_Guide.html#step-3-creating-a-fork-and-testing-your-changes) from the coala project for some details.

**Comment by The-Compiler** [comment]
Looks great now, thank you!

There's one thing I don't get, though - this doesn't seem to be true anymore:

> - Right now, doing `/foo` and pressing enter jumps to the second match. I guess we should just ignore a `/foo` command if incremental search is done?

But I have no idea how you fixed it, or why it's not the case anymore. Can you enlighten me? 😆

**Comment by jupart** [comment]
I know, I checked on that too and noticed it doesn't happen. I'm not sure what the issue was before, but currently (without incremental search), searching the same term twice still gets you the first term on the page. Thanks for bearing with me on my first PR for qb!
